### PR TITLE
feat: unsupported platform warning

### LINKED ISSUE #335: Better handling of unsupported commands
The following tests are not supported on every hardware platform:
```
VerifyAdverseDrops
VerifyUnifiedForwardingTableMode
VerifyTcamProfile
VerifyVxlanConfigSanity
```
Better handling needs to be implemented instead of giving an error. Skip maybe?

### LINKED ISSUE #421: Bug: Exception object not present in TestResult
The exception object is not set in the error attribute of TestResult when an exception occurs in `_collect`.
```
  {
    "name": "DC1-LEAF1A",
    "test": "VerifyAdverseDrops",
    "categories": [
      "hardware"
    ],
    "description": "Verifies there are no adverse drops on DCS7280E and DCS7500E",
    "result": "error",
    "messages": [
      "show hardware counter drop has failed: EapiCommandError (CLI command 2 of 2 'show hardware counter drop' failed: invalid command)"
    ],
    "error": "None", <---------- Here
    "custom_field": "None"
  },
  ```

It could be added in the `wrapper` function of `anta_test` when we check `self.failed_commands`.

### PR DESCRIPTION
**Pull request by mtache**
# Description

If a command defined in a specific test returns the "not supported on this hardware platform" error, the test will be skipped (instead of being in error) and a clear warning will request the user to open a GitHub issue because the test needs to be skipped on specific hardware platform but this has not been implemented.

Example:
```
WARNING  [anta.tests.hardware.VerifyAdverseDrops] Test VerifyAdverseDrops has been skipped because it is not supported on cEOSLab, please open a GitHub issue.
```

Changes:
- Introduce supports() method for AntaDevice.

Fixes #335
Fixes #421 

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have run pre-commit for code linting and typing (`pre-commit run`)
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes (`tox -e testenv`)
- [x] Open PR at https://github.com/jeremyschulman/aio-eapi -> https://github.com/jeremyschulman/aio-eapi/pull/13

### CODE REVIEW

File: anta/device.py [hunk 1]
```diff
@@ -163,6 +250,12 @@ async def collect_commands(self, commands: list[AntaCommand]) -> None:
         """
         await asyncio.gather(*(self.collect(command=command) for command in commands))
 
+    @staticmethod
+    @abstractmethod
+    def is_supported(command: AntaCommand) -> bool:
```
Comment (gmuloc) [inline anta/device.py:255]: so the call would be `device.is_supported(command)` maybe let's change the name in `device.supports_command(command)` ?

File: tests/units/test_device.py [hunk 2]
```diff
@@ -102,6 +102,40 @@ def test__init__(self, device_data: dict[str, Any]) -> None:
...
+            "enable_password": None,
+            "disable_cache": True,
+            "port": None,
+            "ssh_port": None,
+            "tags": None,
+            "timeout": None,
+            "insecure": False,
+            "proto": None,
+        }
+        kwargs = {k: v for k, v in DATA.items() if v is not None and k not in ["host", "username", "password"]}
```
Comment (gmuloc) [inline tests/units/test_device.py:124]: why not write kwargs directly ?

File: tests/units/test_models.py [hunk 3]
```diff
@@ -32,6 +32,19 @@ def test(self) -> None:
         self.result.is_success()
 
 
+class FakeTestWithFailedCommand(AntaTest):
```
Comment (gmuloc) [inline tests/units/test_models.py:35]: can we add a test for the not supported on this hardware result as well?

File: anta/device.py [hunk 4]
```diff
@@ -28,6 +28,93 @@
...
+    def __str__(self) -> str:
+        """returns the error message associated with the exception"""
+        return self.errmsg
+
+
+aioeapi.EapiCommandError = EapiCommandError
+
+
+# patching aioeapi.Device.jsonrpc_exec to test some aio-eapi enhancement
+async def jsonrpc_exec(self, jsonrpc: dict) -> List[Union[Dict, AnyStr]]:  # type: ignore
```
[the remark this answers is not in the document]
Comment (mtache) [reply anta/device.py:62]: Changes need to be made in this PR: https://github.com/jeremyschulman/aio-eapi/pull/13
The type ignore is not relevant for our project.

File: anta/device.py [hunk 5]
```diff
@@ -28,6 +28,93 @@
...
+
+    def __str__(self) -> str:
+        """returns the error message associated with the exception"""
+        return self.errmsg
+
+
+aioeapi.EapiCommandError = EapiCommandError
+
+
+# patching aioeapi.Device.jsonrpc_exec to test some aio-eapi enhancement
```
[the remark this answers is not in the document]
Comment (mtache) [reply anta/device.py:61]: Not released yet

File: tests/lib/fixture.py [hunk 6]
```diff
@@ -9,32 +9,31 @@
...
     mock.username = "toto"
     mock.password = "mysuperdupersecret"
     mock.enable_password = "mysuperduperenablesecret"
-    mock.session = create_autospec(Device)
+    mock.supports = AsyncEOSDevice.supports
     mock.is_online = True
     mock.established = True
-    mock.hw_model = hw_model
+    mock.hw_model = "mocked_AsyncEOSDevice"
     return mock
```
Comment (gmuloc) [inline tests/lib/fixture.py:37]: maybe just make another mocked device instead of reusing?

File: tests/units/test_device.py [hunk 7]
```diff
@@ -102,6 +103,32 @@ def test__init__(self, device_data: dict[str, Any]) -> None:
...
+        Test if the supports() static method parses correctly the aioeapi.EapiCommandError exception
+        """
+        DATA: dict[str, Any] = {"host": "42.42.42.42", "username": "anta", "password": "anta"}
+        kwargs: dict[str, Any] = {
+            "name": "test.anta.ninja",
+            "enable": False,
+            "disable_cache": True,
+            "insecure": False,
+        }
+        device = AsyncEOSDevice(DATA["host"], DATA["username"], DATA["password"], **kwargs)
```
Comment (gmuloc) [inline tests/units/test_device.py:117]: why not using a fixture / mocked_device?

File: pyproject.toml [hunk 8]
```diff
@@ -161,8 +161,6 @@ strict_optional = true
 # for strict mypy: (this is the tricky one :-))
 disallow_untyped_defs = true
 
-mypy_path = "anta"
-
```
Comment (gmuloc) [inline pyproject.toml:165]: is it removed on purpose?

File: anta/models.py [hunk 9]
```diff
@@ -380,7 +380,7 @@ def _init_commands(self, eos_data: Optional[list[dict[Any, Any] | str]]) -> None
         if self.__class__.commands:
             for cmd in self.__class__.commands:
                 if isinstance(cmd, AntaCommand):
-                    self.instance_commands.append(deepcopy(cmd))
+                    self.instance_commands.append(copy(cmd))
```
Comment (gmuloc) [inline anta/models.py:383]: is copy enough for what we are copying?

File: tests/lib/fixture.py [hunk 10]
```diff
@@ -5,41 +5,63 @@
...
+    kwargs = {"name": DEVICE_NAME, "hw_model": DEVICE_HW_MODEL}
+
+    if hasattr(request, "param"):
+        # Fixture is parametrized indirectly
+        kwargs.update(request.param)
+    with patch.object(AntaDevice, "__abstractmethods__", set()):
+        with patch("anta.device.AntaDevice._collect", side_effect=_collect):
+            hw_model = kwargs.pop("hw_model")
+            dev = AntaDevice(**kwargs)  # type: ignore[abstract, arg-type]  # pylint: disable=abstract-class-instantiated, unexpected-keyword-arg
+            dev.hw_model = hw_model
```
Comment (gmuloc) [inline tests/lib/fixture.py:44]: Why pop to add it back?

File: tests/lib/fixture.py [hunk 11]
```diff
@@ -5,41 +5,63 @@
...
 from anta.result_manager.models import TestResult
 from tests.lib.utils import default_anta_env
 
+DEVICE_HW_MODEL = "pytest"
+DEVICE_NAME = "pytest"
+COMMAND_OUTPUT = "retrieved"
+
+
+@pytest.fixture
+def device(request: pytest.FixtureRequest) -> Iterator[AntaDevice]:
```
Comment (gmuloc) [inline tests/lib/fixture.py:27]: I don't know if we should not use a different name for the fixture(s) to make it clear what this is

File: tests/units/result_manager/test_models.py [hunk 12]
```diff
@@ -8,18 +8,18 @@
 
 import pytest
 
-from anta.result_manager.models import TestResult
+# Import as Result to avoid pytest collection
+from anta.result_manager.models import TestResult as Result
```
Comment (gmuloc) [inline tests/units/result_manager/test_models.py:12]: nice!

File: tests/units/test_device.py (hunk 7 above)
Comment (gmuloc) [reply tests/units/test_device.py:116]: does it still need to be tested at the AsyncEOSDevice level or could it be done at abstract AntaDevice one?

File: tests/units/test_device.py [hunk 13]
```diff
@@ -60,188 +51,727 @@
...
+            "host": "42.42.42.42",
+            "username": "anta",
+            "password": "anta",
+            "name": "test.anta.ninja",
+            "insecure": True,
+        },
+        "expected": {"name": "test.anta.ninja"},
+    },
+]
+EQUQLITY_DATA: list[dict[str, Any]] = [
```
Comment (gmuloc) [inline tests/units/test_device.py:70]:
```suggestion
EQUALITY_DATA: list[dict[str, Any]] = [
```

**Review by gmuloc — APPROVED** [review]
🚢 
:shipit:

**Pull request merged**